pyracf/common/utilities/logger.py

```diff
@@ -200,30 +271,49 @@ def redact_result_xml(
         self,
         security_result: Union[str, bytes, List[int]],
         secret_traits: dict,
+        racf_key_map: dict = {},
+        racf_message_key_map: dict = {},
     ) -> str:
         """
         Redacts a list of specific secret traits in a result xml string.
-        Based on the following RACF command pattern:
+        Based on the following RACF command patterns:
             'TRAIT (value)'
+            'TRAIT (subtrait1(value) subtrait2(value))
+            "TRAIT ('value')"
         This function also accounts for varied amounts of whitespace in the pattern.
         """
         if isinstance(security_result, list):
             return security_result
         for xml_key in secret_traits.values():
             racf_key = xml_key.split(":")[1] if ":" in xml_key else xml_key
-            end_pattern = ")"
+            if racf_key in racf_key_map:
+                print(racf_key_map[racf_key])
+                racf_key = racf_key_map[racf_key]
             if isinstance(security_result, bytes):
                 match = re.search(
-                    rf"{racf_key.upper()} +\(", security_result.decode("cp1047")
+                    rf"{racf_key.upper()}( +){{0,}}\(", security_result.decode("cp1047")
                 )
-                end_pattern = end_pattern.encode("cp1047")
             else:
-                match = re.search(rf"{racf_key.upper()} +\(", security_result)
+                match = re.search(rf"{racf_key.upper()}( +){{0,}}\(", security_result)
             if not match:
                 continue
-            security_result = self.__redact_string(
-                security_result, match.end(), end_pattern
-            )
+            security_result = self.__redact_string(security_result, racf_key)
+            if racf_key in racf_message_key_map:
+                racf_key = racf_message_key_map[racf_key]
+            if isinstance(security_result, bytes):
+                security_result = re.sub(
+                    rf"<message>([A-Z]*[0-9]*[A-Z]) [^<>]*{racf_key.upper()}[^<>]*<\/message>",
+                    rf"<message>REDACTED MESSAGE CONCERNING {racf_key.upper()}, "
+                    + r"REVIEW DOCUMENTATION OF \1 FOR MORE INFORMATION</message>",
+                    security_result.decode("cp1047"),
+                ).encode("cp1047")
+            else:
+                security_result = re.sub(
+                    rf"<message>([A-Z]*[0-9]*[A-Z]) [^<>]*{racf_key.upper()}[^<>]*<\/message>",
+                    rf"<message>REDACTED MESSAGE CONCERNING {racf_key.upper()}, "
+                    + r"REVIEW DOCUMENTATION OF \1 FOR MORE INFORMATION</message>",
```

Comment:
Again, may make sense to initialize repeated regex at the beginning of the function:

```python
racf_command_argument_regex = rf"{racf_key.upper()}( +){{0,}}\("
racf_message_regex = rf"<message>([A-Z]*[0-9]*[A-Z]) [^<>]*{racf_key.upper()}[^<>]*<\/message>"
racf_redacted_message_regex = (
    rf"<message>REDACTED MESSAGE CONCERNING {racf_key.upper()}, "
    + r"REVIEW DOCUMENTATION OF \1 FOR MORE INFORMATION</message>"
)
```

Reply:
Yup, all true, but still good to build these once and re-use them.